### PR TITLE
Reports files without Express routes in controllers

### LINKED ISSUE #441: Consider scanning for the existence of Express routes in file before loading it as a controller
An extension of #122

### PR DESCRIPTION
**Pull request by dishad**
Closes #441

### CODE REVIEW

File: lib/checkRoutes.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    if (!content.match(/\.(route|status|send)*?\(([^)]*)\)/)) {
+      badControllers.push(controller)
+    }
+  }
+
+  // Report bad controller files
+  if (badControllers.length > 0) {
+    badControllers[0] = '       - ' + badControllers[0]
+    badControllers = badControllers.join('\n       - ')
+    logger.log('⚠️', `There are files without Express routes located in your controllers directory.\n${badControllers}`.yellow)
```
Comment (kethinov) [inline lib/checkRoutes.js]:
```suggestion
    logger.warn(`There are files without Express routes located in your controllers directory.\n${badControllers}`.yellow)
```

Comment (dishad) [reply lib/checkRoutes.js]: Current PR should reflect these changes

**Comment by kethinov** [comment]
Looks like a coverage drop.

**Comment by dishad** [comment]
Should now achieve 100% coverage

File: lib/checkRoutes.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    // Get all controller files
+    controllerFiles = fs.readdirSync(controllersPath)
+
+    // Loop through files and their content
+    for (let controller of controllerFiles) {
+      let file = path.join(controllersPath, controller)
+      let content = fs.readFileSync(file, 'utf8')
+
+      // If no express routes are found (error files may not have an app.route())
+      if (!content.match(/\.(route|status|send)*?\(([^)]*)\)/)) {
```
Comment (kethinov) [inline lib/checkRoutes.js]: i think there might be some false negatives here

take a look at the [various ways to write express routes](https://expressjs.com/en/guide/routing.html)

wouldn't this test fail something like this?

```javascript
  app.get('/stuff', (req, res) => {
    res.render('something', {})
  })
```

File: lib/checkRoutes.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  const controllersPath = app.get('controllersPath')
+  let badControllers = []
+  let controllerFiles
+
+  // Make sure controllersPath is a directory
+  try {
+    fs.accessSync(controllersPath)
+
+    // Get all controller files
+    controllerFiles = fs.readdirSync(controllersPath)
```
Comment (kethinov) [inline lib/checkRoutes.js]: needs to be recursive

File: lib/checkRoutes.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    // Loop through files and their content
+    for (let controller of controllerFiles) {
+      let file = path.join(controllersPath, controller)
+      let content = fs.readFileSync(file, 'utf8')
+
+      // If no express routes are found (error files may not have an app.route())
+      if (!content.match(/\.(render|status|send)*?\(([^)]*)\)/)) {
```
Comment (kethinov) [inline lib/checkRoutes.js]: this doesn't handle scenarios like this:

```js
// 404.js
const notFound = require('controllers/notFound')

module.exports = app => {
  app.all('*', (req, res) => {
    notFound(app, req, res)
  })
}
```

**Comment by kethinov** [comment]
While testing this I've seen some valid use cases for writing controllers without routes in them. I'm gonna close this PR for now.

**Pull request closed without merging**